test_competitor/src/test_competitor.cpp

```diff
@@ -812,6 +899,114 @@ bool TestCompetitor::FloorRobotPickBinPart(ariac_msgs::msg::Part part_to_pick)
 }
 
+bool TestCompetitor::FloorRobotPickConveyorPart(ariac_msgs::msg::Part part_to_pick)
+{
+  RCLCPP_INFO_STREAM(get_logger(), "Attempting to pick a " << part_colors_[part_to_pick.color] << " " << part_types_[part_to_pick.type] << " from the conveyor");
+
+  bool found_part = false;
+  geometry_msgs::msg::Pose part_pose;
+  rclcpp::Duration elapsed_time(0, 0);
+  builtin_interfaces::msg::Duration time_to_pick;
+  rclcpp::Time detection_time;
+
+  // Change gripper at Kitting Tray Station 2
+  if (floor_gripper_state_.type != "part_gripper")
+  {
+    floor_robot_.setJointValueTarget(floor_kts2_js_);
+    FloorRobotMovetoTarget();
+    FloorRobotChangeGripper("kts2", "parts");
+  }
+
+  // Move robot to predefined pick location
+  floor_robot_.setJointValueTarget(floor_conveyor_js_);
+  FloorRobotMovetoTarget();
+
+  // Find the requested part on the conveyor
+  do {
+      { 
+      // Lock conveyor_parts_ mutex
+      std::lock_guard<std::mutex> lock(conveyor_parts_mutex);
+      auto it = conveyor_parts_.begin();
+      for (; it != conveyor_parts_.end(); ) {
+          auto part = it->first.part;
+          if (part.type == part_to_pick.type && part.color == part_to_pick.color)
+          {
+            part_pose = MultiplyPose(conveyor_camera_pose_, it->first.pose);
+            detection_time = it->second;
+
+            elapsed_time = rclcpp::Time(now()) - detection_time;
+            auto current_part_position_ = part_pose.position.y - (elapsed_time.seconds() * conveyor_speed_);
+            // Check if part hasn't passed the pick location
+            if (current_part_position_ > 0)
+            {
+              time_to_pick.sec = current_part_position_ / conveyor_speed_;
+              // time_to_pick.nanosec = (current_part_position_ / conveyor_speed_ - time_to_pick.sec) * 1e9;
+              // Check if part has more than 5 seconds to arrive at pick location
+              if (time_to_pick.sec > 5.0)
+              {
+                found_part = true;
+                conveyor_parts_.erase(it);
+                break;
+              }
+            }
+            it = conveyor_parts_.erase(it);
+          }
+          else{
+            ++it;
+          }
+        }
+        RCLCPP_INFO_THROTTLE(get_logger(), *get_clock(), 20000, "Waiting for %s %s to arrive on the conveyor", part_colors_[part_to_pick.color].c_str(), part_types_[part_to_pick.type].c_str());
+      } // End lock_guard scope
+    } while (!found_part);
+
+  // Correct robot position to account for part offset on conveyor
+  double part_rotation = GetYaw(part_pose);
+  geometry_msgs::msg::Pose robot_pose = floor_robot_.getCurrentPose().pose;
+  part_pose.position.y = robot_pose.position.y;
+  
+  std::vector<geometry_msgs::msg::Pose> waypoints;
+  waypoints.push_back(BuildPose(part_pose.position.x, part_pose.position.y,
+                                part_pose.position.z + 0.15, SetRobotOrientation(part_rotation)));
+  FloorRobotMoveCartesian(waypoints, 1, 1);
```

Comment:
After this move is completed we should check again to make sure the part has not passed the pick position

Reply:
Implemented in comit [4c8c8d6](https://github.com/usnistgov/ARIAC/pull/262/commits/4c8c8d6688d4db90e4d75ed890c9e16005a66497)